### PR TITLE
[3.0] Refactor MetricsConfig

### PR DESCRIPTION
**Pull request by kevinw66**
## What is the purpose of the change

1.Refactor MetricsConfig
2.Add dubbo-metrics module
3.Support nested class configuration resolver

### CODE REVIEW

File: dubbo-config/dubbo-config-spring/src/test/resources/org/apache/dubbo/config/spring/metrics-prometheus.xml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+       xmlns:dubbo="http://dubbo.apache.org/schema/dubbo"
+       xmlns="http://www.springframework.org/schema/beans"
+       xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-4.3.xsd
+    http://dubbo.apache.org/schema/dubbo http://dubbo.apache.org/schema/dubbo/dubbo.xsd
+    ">
+
+    <!-- current application configuration -->
+    <dubbo:application name="demo-consumer" />
+
+    <dubbo:metrics protocol="prometheus">
```
Comment (kylixs) [inline dubbo-config/dubbo-config-spring/src/test/resources/org/apache/dubbo/config/spring/metrics-prometheus.xml:27]: Should we remove the `<dubbo:prometheus> ... <dubbo:prometheus/>` ?
```xml
<dubbo:metrics protocol="prometheus">
    <dubbo:prometheus-exporter enabled="true" enable-http-service-discovery="true" http-service-discovery-url="localhost:8080" metrics-port="20888" metrics-path="/metrics" />
    <dubbo:prometheus-pushgateway enabled="true" base-url="localhost:9091" push-interval="30" username="username" password="password" job="job" />
</dubbo:metrics>
```

File: dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/MetricsFactory.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+public interface MetricsFactory {
+
+    /**
+     * Create metrics.
+     *
+     * @param url
+     * @return metrics
+     */
+    @Adaptive("protocol")
+    Metrics getMetrics(URL url);
```
Comment (kylixs) [inline dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/MetricsFactory.java:37]: The name of `Metrics` / `MetricsFactory` is not very appropriate, here is a suggestion: `MetricsProvider` / `MetricsProviderFactory`

File: dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/Metrics.java [hunk 3]
```diff
@@ -14,16 +14,15 @@
...
-    @Override
-    public Class<?> getService() {
-        return MetricsService.class;
-    }
+/**
+ * Monitor. (SPI, Prototype, ThreadSafe)
+ *
+ * @see org.apache.dubbo.metrics.MetricsFactory#getMetrics(org.apache.dubbo.common.URL)
+ */
+public interface Metrics extends Node, MetricsService {
```
Comment (kylixs) [inline dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/Metrics.java:26]: `MetricsProvider` should not implement `MetricsService`. I think `MetricsService` is a service that provides metrics query, not implemented by metrics provider.

File: dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/MetricsFactory.java (hunk 2 above)
Comment (kylixs) [reply dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/MetricsFactory.java:37]: MetricsProvider also seems inappropriate. What capabilities can Prometheus' SPI provide? Think about this interface

File: dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/Metrics.java (hunk 3 above)
Comment (kevinw66) [reply dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/Metrics.java]: deleted

File: dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/MetricsFactory.java (hunk 2 above)
Comment (kevinw66) [reply dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/MetricsFactory.java]: Currently deleted. Wait for next pull request to complete the implementation

File: dubbo-config/dubbo-config-spring/src/test/resources/org/apache/dubbo/config/spring/metrics-prometheus.xml (hunk 1 above)
Comment (kevinw66) [reply dubbo-config/dubbo-config-spring/src/test/resources/org/apache/dubbo/config/spring/metrics-prometheus.xml:27]: removed

**Review by kylixs — APPROVED** [review]
Looks good.

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/MethodUtils.java [hunk 4]
```diff
@@ -395,4 +396,35 @@ static Method findOverriddenMethod(Method overrider, Class<?> declaringClass) {
...
     }
+
+    /**
+     * Extract fieldName from set/get/is method. if it's not a set/get/is method, return empty string.
+     * If method equals get/is/getClass/getObject, also return empty string.
+     *
+     * @param method method
+     * @return fieldName
+     */
+    static String extractFieldName(Method method) {
```
Comment (CrazyHZM) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/MethodUtils.java:407]: Missing unit test, same below.

File: dubbo-common/src/test/java/org/apache/dubbo/config/context/ConfigManagerTest.java [hunk 5]
```diff
@@ -126,6 +126,7 @@ public void tesModuleConfig() {
     @Test
     public void tesMetricsConfig() {
         MetricsConfig config = new MetricsConfig();
+        config.setProtocol("prometheus");
```
Comment (CrazyHZM) [inline dubbo-common/src/test/java/org/apache/dubbo/config/context/ConfigManagerTest.java:129]: Replace with `PROTOCOL_PROMETHEUS`.
Same below.

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/MethodUtils.java (hunk 4 above)
Comment (kevinw66) [reply dubbo-common/src/main/java/org/apache/dubbo/common/utils/MethodUtils.java:407]: done

File: dubbo-common/src/test/java/org/apache/dubbo/config/context/ConfigManagerTest.java (hunk 5 above)
Comment (kevinw66) [reply dubbo-common/src/test/java/org/apache/dubbo/config/context/ConfigManagerTest.java]: done

**Review by CrazyHZM — APPROVED** [review]
LGTM

**Pull request merged**